core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -280,10 +280,54 @@ private[spark] class TaskSchedulerImpl(
         }
       }
     }
+    if (!launchedTask && isTaskSetCompletelyBlacklisted(taskSet)) {
+      taskSet.abort(s"Aborting ${taskSet.taskSet} because it has a task which cannot be scheduled" +
+        s" on any executor due to blacklists.")
+    }
     return launchedTask
   }
 
   /**
+   * Check whether the given task set has been blacklisted to the point that it can't run anywhere.
+   *
+   * It is possible that this taskset has become impossible to schedule *anywhere* due to the
+   * blacklist.  The most common scenario would be if there are fewer executors than
+   * spark.task.maxFailures. We need to detect this so we can fail the task set, otherwise the job
+   * will hang.
+   *
+   * The check here is a balance between being sure to catch the issue, but not wasting
+   * too much time inside the scheduling loop.  Just check if the last task is schedulable
+   * on any of the available executors.  So this is O(numExecutors) worst-case, but it'll
+   * really be fast unless you've got a bunch of things blacklisted.  Its possible it won't detect
+   * the unschedulable task immediately, but if it returns false, there is at least *some* task
+   * that is schedulable, and after scheduling all of those, we'll eventually find the unschedulable
+   * task.
+   */
+  private[scheduler] def isTaskSetCompletelyBlacklisted(
```

Comment:
I think it would be cleaner to add this method to the TaskSetManager class  (and then you don't need the pollPendingTask method) -- and then just pass in the executorsByHost map.  That also makes things a little easier to change in the future, if there gets to be some easier way of checking if a particular task set is completely blacklisted.